### ISSUE TITLE
No indication there are unsynced quests if uploading automatically but no WiFi or data connection

### ISSUE THREAD

**Issue by peternewman** [streetcomplete/StreetComplete#3618]
No indication there are unsynced quests if uploading automatically but no WiFi or data connection.

I found this when I'd done a full reset of SC and had mobile data turned off and no WiFi connection at the time.

I'd suggest the space taken up by the manual upload arrow in Off/Only on WiFi should just have a counter for outstanding quests.

**How to Reproduce**

1. "Upload answers automatically" set to On.
2. Airplane mode turned on (or other situation where no WiFi and no data connection
3. Solve a quest
4. Aside from the Undo symbol, there's no indication there are unsynced quests, and certainly no indication of how many.

**Versions affected**
v38.2

**Comment by westnordost** [comment]
Why?

**Comment by smichel17** [comment]
He had pending changes, and did not realize. Then he wiped the SC data and lost his pending changes.

**Comment by westnordost** [comment]
Why did you do a full reset?

**Comment by smichel17** [comment]
If it is implemented, I would suggest, instead of taking up so much space, show a small indicator next to the star count, along these lines (but probably not exactly like this):
![image](https://user-images.githubusercontent.com/6563664/148258327-fdd4fb3a-4b88-4e73-912a-2bbbd06474d9.png)

**Comment by peternewman** [comment]
> Then he wiped the SC data and lost his pending changes.

Actually it was the other way round, I wiped all of SC, which reset all my settings.

> He had pending changes, and did not realize.

Yes, I then noticed there was no indication of unsynced quests if you're in auto-upload (even if it's not succeeding)

I didn't thankfully lose any data, but I wouldn't have known if I had, or that uploading wasn't working.

To put it another way, you can get in the same state as "Only on Wi-Fi" by being in On but with data disabled, so why do they indicate differently?

> Why did you do a full reset?

To try and get a handle on my disk space again, it's currently at 125MB of user data and 40MB of cache, but it was a lot higher than that. I'd been in two other countries, including mapping along a multi-hour train journey through one of them, and I'm unlike to return to either country any time soon, so I was trying to purge that space sooner, rather than needing to wait a week for it to expire.

It felt like #3475 hadn't worked, but maybe I was being impatient, my user data was still very large anyway (although will it ever clear down things like ignored quests)?

> If it is implemented, I would suggest, instead of taking up so much space, show a small indicator next to the star count, along these lines (but probably not exactly like this): ![image](https://user-images.githubusercontent.com/6563664/148258327-fdd4fb3a-4b88-4e73-912a-2bbbd06474d9.png)

That's odd, for me (SC 38.2/Android 9), I always have the gap where upload manually/Only on Wi-Fi counter would be, mine never squash together like that, unless this is a mockup? Hence suggesting using the existing dead space to have just a total.

**Comment by smichel17** [comment]
It's a mockup, but all I added is the orange (+). The base screenshot was from the repo readme.

**Comment by westnordost** [comment]
>>  Why did you do a full reset?
> 
> To try and get a handle on my disk space again

Why didn't you select "Delete Cache" in the app's settings?

**Comment by peternewman** [comment]
> The base screenshot was from the repo readme.

I'd be curious if that matches someone else's phone in auto-upload, or maybe it's just an old screenshot.

> ~Why didn't you select "Delete Cache" in the app's settings?~
> 
> Nevermind, you wrote that you thought it didn't work.

I think I'm right in thinking it will only ever clear some stuff won't it @westnordost ?

**Comment by westnordost** [comment]
> I think I'm right in thinking it will only ever clear some stuff won't it @westnordost ?

What do you mean with "some stuff"? It won't clear your quest answers, your statistics, your login. But it clears **all** downloaded map data, all quests, all downloaded map tiles.

**Comment by peternewman** [comment]
> > I think I'm right in thinking it will only ever clear some stuff won't it @westnordost ?
> 
> What do you mean with "some stuff"? It won't clear your quest answers, your statistics, your login. But it clears **all** downloaded map data, all quests, all downloaded map tiles.

Stuff like hidden quests etc. When you say quest answers, do you mean ones that haven't been uploaded yet, or all my historical ones? Because I really don't care about the latter.

I've taken the bold step of hiding most of this discussion because it's off-topic to the reported issue; I just happened to mention it in passing to explain how I got to the state where I found the bug, as someone who's frequently asked for things to be Wi-Fi orientated, I thought I should give some background.

**Comment by westnordost** [comment]
As I said, it only clears the downloaded stuff: Downloaded map data, downloaded notes, downloaded map tiles. The setting isn't named like "Delete Cache" for no reason.

It does not clear which quests you've hidden, it does not clear the quest answers you gave and not any of your settings.

**Comment by peternewman** [comment]
> > I got to the state where I found the bug
> 
> Well it's not a bug.

Maybe this should be a feature request then, but there's different behaviour between two different auto-upload mechanisms, which seems unexpected to me.

Aside from a lack of radio signal, my phone's radio module or OS seems a bit buggy and it sometimes gets stuck with no signal/emergency calls only, so quests wouldn't be able to upload in either of those cases either.

**Comment by westnordost** [comment]
> but there's different behaviour between two different auto-upload mechanism

No, not really. What are you referring to?

**Comment by westnordost** [comment]
And how does the upload only in wifi mode and the auto-sync mode then differ? Airplane mode turns off wifi, too.

**Comment by peternewman** [comment]
Yeah that was only an easy reproduction. Otherwise leave the phone in it's normal mode and walk up a hill/go underground out of mobile and WiFi reception.

**Comment by westnordost** [comment]
Okay, but how exactly does the behavior differ then? Auto-sync off tries to sync on mobile and on wifi connection. Auto sync only on wifi tries to sync only on wifi. Where is the issue?

**Comment by peternewman** [comment]
The pending uploads indication (with the upward facing arrow) and it's counter of quests which haven't been uploaded is only present when in WiFi mode despite it being perfectly possible to have pending uploads in either mode.

**Comment by westnordost** [comment]
Yes,  well, this is deliberate. Users with auto sync on (default) should not be bothered with ui elements they don't need and are not explained anywhere.

**Comment by peternewman** [comment]
So how are they supposed to know if it's not syncing for a variety of reasons? Surely everyone should know about the counter even if they don't need the arrow to force an upload?

**Comment by westnordost** [comment]
Google Docs also does not show a counter of changed things, right? The document is saved automatically. So what does Google Docs do if you are offline / google can't be reached?

**Comment by matkoniecz** [comment]
They show warning popup about failed connection (I have no access to the computer, but it should be fairly easy to test - open Google docs, switch off router).

**Comment by peternewman** [comment]
> Google Docs also does not show a counter of changed things, right?

I'd say that stretched the analogy a little, as each SC edit is standalone, the equivalent would be making changes in multiple Google Docs.

> The document is saved automatically. So what does Google Docs do if you are offline / google can't be reached?

Having just tested by turning off my WiFi (even easier), on my laptop in the web browser, I managed to type another word, then afterwards it popped up a nearly modal dialogue about trying to connect and wouldn't let me do anything apart from copy the text out of it.

I'd also point out Google isn't the pinnacle of offline working so not necessarily what we should aim for. In the past I've made offline changes on a spreadsheet listing things, while someone else made online ones, and it overwrote the contents of the cell when my changes uploaded. I think at least we'll get a conflict warning instead with SC.

**Comment by matkoniecz** [comment]
Maybe, maybe showing upload counter would be a good idea if changes were not uploaded for some longer time or amount of waiting changes is high?

For reference: Google Docs completely block editing (presumably, to avoid edit conflicts and unsure that Google ads can continue to work in other tabs)

![screen01](https://user-images.githubusercontent.com/899988/155880103-64dabadd-7169-4954-99dc-3b086ad8d60b.png)

**Comment by peternewman** [comment]
> Maybe, maybe showing upload counter would be a good idea if changes were not uploaded for some longer time or amount of waiting changes is high?

To me immediately seems sense, but otherwise waiting for a time seems better than a high number of changes. Surely every edit is valuable as someone has taken the time to do it? Although I appreciate mobile Internet can be variable so perhaps it should wait for a period of time before showing the counter, but it would need to be fairly low, otherwise there's a risk you make an edit and close the app before the timer has elapsed and don't know your edits haven't uploaded.

**Comment by mnalis** [comment]
Maybe show an icon only where there are outstanding quests (i.e. quests which have not  yet been successfully uploaded)?

So basically that would in normal circumstances for users having auto-upload enabled only briefly show the icon, e.g.:
- you would solve an quest
- icon would immediately show with "1"
- background task would start uploading
- when it finished uploading (a half a second or whatever later) the icon would disappear completely again. 


That would provide more information in case of upload problems without using useful screen space permanently, but it might possibly be seen as annoying flicker too. Note if you add delay of few seconds to avoid that flicker, you could be back at square one - as user could close the app assuming it has finished uploading, because they didn't see that icon in time.


On the other hand, if the upload was unsuccessful, the icon would remain on screen (with count increasing as they solved more quests), indicating to a user that there is a problem.
